### PR TITLE
Cleanup Animation frame_formats.

### PR DESCRIPTION
**Pull request by anntzer**
- Warn if the requested frame format is unsupported and fallback occurs.
- FFMpegFileWriter does not support 'raw' or 'rgba' (AFAICT there's no
  way to forward the relevant frame size/bitdepth info).
- ImageMagickFileWriter supports 'rgba' but requires explicit frame_size
  and depth to do so (they are the same as for ImageMagickWriter, and
  passing them doesn't negatively impact other formats), but it does not
  support 'raw' (per https://imagemagick.org/script/formats.php).

Test script:
```python
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation

fig, ax = plt.subplots()

x = np.arange(0, 2*np.pi, 0.01)
line, = ax.plot(x, np.sin(x))

def animate(i):
    line.set_ydata(np.sin(x + i / 50))  # update the data.
    return line,

ani = animation.FuncAnimation(
    fig, animate, interval=50, blit=True, save_count=20)

plt.rcParams["animation.frame_format"] = "rgba"  # or "raw"
writer = animation.FFMpegFileWriter()  # or animation.ImageMagickFileWriter()
ani.save("/tmp/movie.mp4", writer=writer)
```

(Also move import of PIL up, as it's a standard dependency now.)

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/animation.py [hunk 1]
```diff
@@ -296,8 +297,11 @@ def __init__(self, fps=5, codec=None, bitrate=None, extra_args=None,
 
         super().__init__(fps=fps, metadata=metadata, codec=codec,
                          bitrate=bitrate)
-
-        self.frame_format = 'rgba'
+        # The rcParam only applies to FileMovieWriters, but should be read here
+        # because setting to frame_format to 'rgba' would otherwise spuriously
+        # warn that that format is unsupported e.g. for FFMpegFileWriter.
+        self.frame_format = (mpl.rcParams['animation.frame_format']
+                             if isinstance(self, FileMovieWriter) else 'rgba')
```
Comment (dopplershift) [inline lib/matplotlib/animation.py:304]: A base class that checks for a specific subclass? Ewww. There *has* to be a better way to accomplish this. Set `_frame_format` to `None` by default and do this in the getter on `None`?

Comment (anntzer) [reply lib/matplotlib/animation.py:304]: While that's doable, note that the fundamental problem here is that FileMovieWriter should not really inherit from MovieWriter: there's a bunch of methods where the FileMovieWriter override doesn't call the `super()` method (`setup`, `_frame_sink`, `grab_frame`); this PR just shows that likewise there's `__init__` functionality that isn't really fully inherited.  I guess the proper approach may be to move some of the common functionality to AbstractMovieWriter and make both MW and FMW directly inherit from AMW instead?...

**Comment by QuLogic** [comment]
> FFMpegFileWriter does not support 'raw' or 'rgba' (AFAICT there's no
> way to forward the relevant frame size/bitdepth info).

I'm not sure about the class in Matplotlib, but at least ffmpeg can accept rgba24 frames (specifying size, etc. via command-line info) https://ffmpeg.org/ffmpeg-formats.html#rawvideo

**Comment by anntzer** [comment]
I specifically meant the Matplotlib class, when used for reconstituting a movie *from on-disk image (rgba) files*.  If you (or someone else) knows the right invocation to make the example given above work, then that's even better :)

**Comment by QuLogic** [comment]
Yes, I linked to the wrong format, but it should be possible. I just need to test out the other frame formats to be sure I didn't break it.

File: lib/matplotlib/animation.py (hunk 1 above)
Comment (QuLogic) [reply lib/matplotlib/animation.py:304]: I think you can revert this with #19056?

File: lib/matplotlib/animation.py [hunk 2]
```diff
@@ -629,8 +633,7 @@ class FFMpegFileWriter(FFMpegBase, FileMovieWriter):
     Frames are written to temporary files on disk and then stitched
     together at the end.
     """
-    supported_formats = ['png', 'jpeg', 'ppm', 'tiff', 'sgi', 'bmp',
-                         'pbm', 'raw', 'rgba']
+    supported_formats = ['png', 'jpeg', 'ppm', 'tiff', 'sgi', 'bmp', 'pbm']
```
Comment (QuLogic) [inline lib/matplotlib/animation.py:636]: Add 'raw' and 'rgba' back.

File: lib/matplotlib/animation.py [hunk 3]
```diff
@@ -744,14 +747,15 @@ class ImageMagickFileWriter(ImageMagickBase, FileMovieWriter):
 
     Frames are written to temporary files on disk and then stitched
     together at the end.
-
     """
 
-    supported_formats = ['png', 'jpeg', 'ppm', 'tiff', 'sgi', 'bmp',
-                         'pbm', 'raw', 'rgba']
+    supported_formats = ['png', 'jpeg', 'ppm', 'tiff', 'sgi', 'bmp', 'pbm',
+                         'rgba']
```
Comment (QuLogic) [inline lib/matplotlib/animation.py:753]: Isn't 'raw' the same as 'rgba'?

**Review by QuLogic — CHANGES_REQUESTED** [review]
Can you revert the raw/rgba removal bits, and then I can add tests into #19056.

File: lib/matplotlib/animation.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/animation.py:304]: yup

File: lib/matplotlib/animation.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/animation.py:636]: yup

File: lib/matplotlib/animation.py (hunk 3 above)
Comment (anntzer) [reply lib/matplotlib/animation.py:753]: not from the PoV of ImageMagick, but I changed that PR to make instead force ImageMagick to interpret the files as rgba.

**Review by QuLogic — APPROVED** [review]

**Comment by anntzer** [comment]
Actually, the check for FileMovieWriter when setting the frame format to rgba is still needed for HTMLWriter, which does not support rgba...

**Comment by QuLogic** [comment]
Isn't that overwritten in `FileMovieWriter.__init__` anyway?

**Comment by anntzer** [comment]
Ah yes, that's a good idea.

**Comment by QuLogic** [comment]
Unfortunately, seems to have broken docs.

**Comment by anntzer** [comment]
hopefully fixed now...

**Review by timhoffm — APPROVED** [review]

**Pull request merged**